### PR TITLE
Ignore nodes with extremely low policy priors

### PR DESCRIPTION
**Pull request by WhiteHalmos**
Skip creating child nodes with <0.001 policy prior during expansion.

### CODE REVIEW

**Comment by WhiteHalmos** [comment]
```
leelaz-next_ed26 v leelaz-trim_ed26 (233/400 games)
board size: 19   komi: 7.5
                   wins              black          white        avg cpu
leelaz-next_ed26     92 39.48%       40  34.19%     52  44.83%    341.48
leelaz-trim_ed26    141 60.52%       64  55.17%     77  65.81%    339.31
                                     104 44.64%     129 55.36%
```
Ringmaster config:
```python
competition_type = 'playoff'
LZ_ARGS = '/projects/lztest/{} --timemanage off --noponder -r10 -g -t1 -d -w {}'
def LZ(vals):
  commands = ['time_settings 450 0 0']
  return Player(LZ_ARGS.format(*vals), startup_gtp_commands=commands)
players = dict()
p1 = 'leelaz-next_ed26'
p2 = 'leelaz-trim-math_ed26'
players[p1] = LZ(p1.split('_'))
players[p2] = LZ(p2.split('_'))
board_size = 19
komi = 7.5
matchups = [
  Matchup(p1, p2, alternating=True, number_of_games=400),
]
```
```
>>> sprt.SPRT(141,92,1,195,225,0.05,0.05)
'H1'
```

**Comment by Eddh** [comment]
Interesting. In theory this should be relevant only in cases where there is a big drop in winrate because of fpu reduction. This is a bigger difference than I expected from this.

File: src/UCTNode.cpp [hunk 1]
```diff
@@ -92,6 +92,7 @@ bool UCTNode::create_children(std::atomic<int>& nodecount,
 
     auto legal_sum = 0.0f;
     for (const auto& node : raw_netlist.first) {
+        if (node.first < 0.001f) continue;
```
Comment (TFiFiE) [inline src/UCTNode.cpp]: As a magic number, this should at the least be captured in a constant or, preferably, be made a configurable parameter (to be exposed by USE_TUNER).

**Comment by killerducky** [comment]
Maybe this is due to the increased speed of e.g. uct_select_child on CPU? But if you run with `-t 2` or higher GPU will be the bottleneck and CPU speed will not matter at all. Can you try that? Also why disable timemanage?

**Comment by killerducky** [comment]
Also I think this should only be done for match games, not self-play. Perhaps use the same trick as timemanage and overload the `-m` switch.

**Comment by Eddh** [comment]
Why shouldn't it be done for self-play ? the noise can easily take care of exploration and those low policy moves are almost never explored at root level anyway.

**Comment by Mardak** [comment]
From my rotation explorations in https://github.com/gcp/leela-zero/issues/1007#issuecomment-373833483 and https://github.com/gcp/leela-zero/issues/877#issuecomment-373255981, there have often been the correct capture/punish-LZ move that had N < 0.1%.

For example from the self-atari issue, here's the L11 move 120 to capture LZ with 1000 visits for each of the 8 rotations with 5b90bd32:
```
seed 22: L11 ->     472 (V: 100.00%) (N:  0.05%) PV: L11 R8 F9 F10 E9 E10 D10 D11 C11 D9 C10
seed:  6 L11 ->     483 (V: 100.00%) (N:  0.05%) PV: L11 R8 F9 F10 E9 E10 D10 D11 D9 C11 E11
seed:  2 L11 ->     491 (V: 100.00%) (N:  0.15%) PV: L11 R8 F9 F10 E9 E10 D10 D11 C11 D9 C10 G8
seed: 11 L11 ->     448 (V: 100.00%) (N:  0.08%) PV: L11 R8 F9 E11 D9 C11 F11 F10 E10 G9 F8 G8
seed:  3 L11 ->     497 (V: 100.00%) (N:  0.15%) PV: L11 R8 F9 E11 D10 D11 C11 C10 C9 B10
seed:  4 L11 ->     496 (V: 100.00%) (N:  0.17%) PV: L11 R8 F9 F10 E9 E10 D10 D11 D9 C11 E11 E12
seed: 12 L11 ->     489 (V: 100.00%) (N:  0.13%) PV: L11 R8 F9 F10 E10 E9 F8 E11 D10 D11 F11
seed:  1 L11 ->     490 (V: 100.00%) (N:  0.02%) PV: L11 R8 F9 F10 E10 E9 F8 E11 D10 D11 F11 C10 D9
```

With this PR, half of the possible rotations would no longer have found the capture.

Would that means it would make it harder to learn that this move should have a higher policy prior?

Although at a high level, basically this change magnifies what the net knows or fails to know. So if it doesn't know about self-atari in certain rotations, then it will more likely fail, but also if it thinks a certain line of play is better, it won't waste visits on things it learned to not play anyway.

**Comment by Eddh** [comment]
Oh and the dirichlet noise is applied after so yeah it shouldn't be used in selfplay.

**Comment by WhiteHalmos** [comment]
@Eddh Opps! Updated SPRT test, thanks for catching that. I thought the results looked wonky!

@killerducky Happy to have more people help test this. My machine is tied up right now with the currrent test.

@Mardak Agreed that this is not good for self-play.

File: src/UCTNode.cpp (hunk 1 above)
Comment (WhiteHalmos) [reply src/UCTNode.cpp]: Agreed, but I'd like others to easily test this and for more to weigh in before making this more complicated than necessary. Here's a branch with more math to back it up: https://github.com/WhiteHalmos/leela-zero/commit/1d8ec2d8e0c11d3a57850a00dced1d1b5cba979c

**Comment by WhiteHalmos** [comment]
@petgo3 Assuming LZ is resource constrained on KGS, this very minor change might be worth trying for an extra boost.

**Comment by Eddh** [comment]
What do you think about applying Dirichlet noise before trimming ? Would that allow for using this PR in self-play games ?

**Comment by ihavnoid** [comment]
My general feeling is that the threshold 0.001 seems a bit arbitrary.  If we are doing ten million playouts, for example, there probably is some value to try a couple of moves for the lowest probabilities just in case something is wrong with the network.  On a ten thousand playout setting, skipping anything lower than 0.001 might make sense.

**Comment by WhiteHalmos** [comment]
@ihavnoid See this for a more mathematically sound approach: https://github.com/WhiteHalmos/leela-zero/commit/1d8ec2d8e0c11d3a57850a00dced1d1b5cba979c

**Comment by fell111** [comment]
It's better to introduce a new parameter to set the threshold. Then we can use the same binary for testing. For self play, no threshold will be added. And in matches/other games, we could tune this parameter to get better winrate.

**Comment by gcp** [comment]
I'd like to understand why this makes a difference. We discussed this a lot during #980 and related threads, but it's something I'd have enabled when we surpassed half our memory budget.

If the strength increase is not simply due to the speedup, it means the current UCT behavior is too exploratory.

>It's better to introduce a new parameter to set the threshold. 

Nope. You can do this while tuning this value, but relying on the user to set magic values is incredibly hostile.

**Comment by gcp** [comment]
>@ihavnoid See this for a more mathematically sound approach: WhiteHalmos/leela-zero@1d8ec2d

You can estimate this off of the time control to make it better, but then you still have the critical question: what are you going to do in "infinite analysis"?

You will have problems if you control the tree search by determining these values before starting your search, that's why I think this is better tackled in UCT itself.

**Comment by gcp** [comment]
Hmm, actually...currently MAX_TREE_SIZE is at most 100'000'000, which means the maximum amount of playouts will be around 280k, which means that independent of PUCT, the maximum ratio between best PSA and worst PSA would be 0.0019 in order for us to discover it.

I guess a dynamic formula that caps at no less from that number would work, i.e. it is guaranteed to not make the program worse? But the question is, would you still see the strength improvement you saw now?

**Comment by ihavnoid** [comment]
@gcp - I would argue that the MAX_TREE_SIZE parameter should be command-line controllable so that the user can select reasonable values depending on their setup.  Maybe we can set the cutoff threshold based on whatever user-provided value this is.

If possible I would think that we would want the MAX_TREE_SIZE to be dynamically detected but unfortunately I failed to find any portable method that can do that (e.g., measure the process's memory upper bound in a reliable way)

**Comment by Eddh** [comment]
"If it's mathematically safe, why not prune the root after setting Dirichlet noise?"
Unless I'm misunderstanding something, there is still a potential problem with the first 30 moves, since even 0.0001 policy moves can be picked sometimes. But honestly I think the noise can take care of that well enough and the impact would be very small.

**Comment by gcp** [comment]
>Unless I'm misunderstanding something, there is still a potential problem with the first 30 moves, since even 0.0001 policy moves can be picked sometimes

You are misunderstanding something. It's not possible to select a move that the search did not explore.

**Comment by Eddh** [comment]
Ahh I see nevermind then.

**Comment by gcp** [comment]
>I would argue that the MAX_TREE_SIZE parameter should be command-line controllable so that the user can select reasonable values depending on their setup.

No way. Allowing the user to set the maximum memory usage, that makes more sense, though.

But we are so wasteful with this right now, and this patch will change the behavior a lot, that it you get back to the original problem I raised above. You literally even risk that the program will end up playing WEAKER if you give it MORE memory, because it will drop the value of this magic parameter that doesn't just control memory but also has an effect on the strength.

I repeat: the only reason I'd consider taking it right now is because the *current* code has this limit, so it can't make things worse. But the approach is very deeply problematic because of the above.

**Comment by WhiteHalmos** [comment]
@gcp Agreed with all your points. This PR was more intended to get the idea of improving UCT search through better pruning. I have more math to back up https://github.com/WhiteHalmos/leela-zero/tree/trim_math but it needs lots more testing.

Updated results. Makes me realize that I need to consistently test with 400 games since the winrate kept dropping from 65% in ~100 games to 60% in ~200 games to now 56% in ~400 games.
```
leelaz-next_ed26 v leelaz-trim_ed26 (400/400 games)     
board size: 19   komi: 7.5  
                   wins              black          white        avg cpu                                        
leelaz-next_ed26    176 44.00%       78  39.00%     98  49.00%    341.08                                        
leelaz-trim_ed26    224 56.00%       102 51.00%     122 61.00%    338.98                                        
                                     180 45.00%     220 55.00%
```
Literally one loss away from SPRT significance. Strange that 56% after 400 games isn't SPRT significant on (0, 35).
```python
>>> -400 * math.log(1/.56-1)/math.log(10)
41.89414020800522
>>> sprt.SPRT(224,1,176,0,35,0.05,0.05)
''
>>> sprt.SPRT(224,1,175,0,35,0.05,0.05)
'H1'
```

**Comment by gcp** [comment]
SPRT is stricter because it assumes you will potentially play an infinite amount of games and cherry pick the PASS result.

**Comment by WhiteHalmos** [comment]
Closing this PR as it has achieved its purpose: Demonstrating that there is performance improvements to be had along with huge memory savings by trimming the UCT search tree.

**Comment by gcp** [comment]
I tried to implement a "conservative" version of this, i.e. that gives identical results but saves memory, by abusing MAX_TREE_SIZE, but it's tricky.

Assume we are on the second visit of a node. On the first visit we got a 0% winrate after expanding the first node.

Now, a node with an arbitrarily small policy prior can be selected. It's independent of max_playouts or anything. It does not matter that this node will not be part of the final selection, it still influences in the tree shape and returned score.

**Pull request closed without merging**